Add ERC: Consolidated Dynamic Storage (CDS) (#774)
ethereum/ERCs: ERCS/erc-7844.md

- While centralization of the storage layer seems risky on the surface level, it is actually a responsible approach: with CDS, the potential for data issues is limited to the scope a single contract.  Additionally, simple permissions systems can effectively block all direct access by non-autonomous agents, leaving the burden of risk solely on the connected pure contracts.
- **Collision-Free Storage** ensures logical separation of storage spaces via hashed namespaces and configurable offsets.
- **Optional Batch Operations** significantly reduce gas costs for high-throughput systems by minimizing external calls.
- **State Transparency** provides query functions that allow developers to dynamically inspect storage layouts and interact with stored data.

By solving these limitations, CDS simplifies upgradeability, reduces gas costs, and eliminates storage layout constraints, empowering developers to build evolving systems like DeFi protocols, modular frameworks, and DAOs with greater flexibility and confidence.

![Adaptability](../assets/erc-7844/Adaptability.png)

## Specification 

> **Extendable Structs and Storage Spaces**

Reviewer: Please use subheadings for these (eg `### Extendable Structs and Storage Spaces`.)

However, this raises a problem when implementing extendable structs: *if the last member is a string, we might reference back to it, tricking the system into believing the `stringIndex` is a `bitCount`.* Hence, we leverage `safeIndex`, which simply records the most recent valid index we can use to derive `bitCount`.  Critically, if `safeIndex` is zero, we are still safe from complications, because a zero `strindex` doubles as a valid `bitCount` in that instance.

Example Implementation:
```solidity
function insert_new_member(
	uint256 valType,
	uint256 valSize,
	uint256 storageSpace
) external
{
	verify type and size
	retrieve memberData
	
	if(type not string)
	{
		get safeIndex
		assign bitCount := prev bitCount + sizeof previous
		verify size
		get storage page
		verify we will not overflow
		if overflow, push to next page (update bitCount to head of next page)
		pack memberData
		store memberData
		update safeIndex, members in state data for storage space
	}
	if(type is string)
	{
		get stringIndex
		pack with type
		store packed metadata in memberData
		increment stringIndex, members
		store updated state data for storage sapce
	}

}
```

> ## Interfaces

Reviewer: These should be at level three or deeper (eg. `### Interfaces`)

```solidity
    function init_create(
    uint256[] memory types, 
    uint256[] memory sizes
) external;

    function insert_new_member(
    uint256 valType, 
    uint256 size, 
    uint256 storageSpace
) external;

function push(
	uint256 storageSpace
) external;

function pushMany(
    uint256 amount, 
    uint256 storageSpace
) external;

function put(
    uint256 data, 
    uint256 memberIndex, 
    uint256 entryIndex, 
    uint256 storageSpace
) external;

function put_string(
	string memory data, 
	uint256 memberIndex, 
	uint256 entryIndex, 
	uint256 storageSpace
) external;

> (optional) function put_batch(

Reviewer: I'd recommend putting optional functions into separate interfaces for clarity.

- **Proven Feasibility**: Working, robust models of CDS already exist.  The EVM’s mature auditing ecosystem ensures that well-designed implementations can be secure and reliable.
- **Launchpad for Plug-and-Play Infra Solutions**  CDS’s centralized storage layer, paired with its low-level access, forms a robust foundation for further system-level enhancements.  Frameworks like [HoneyBadger](https://honeybadgerframework.com) demonstrate CDS’s real-world feasibility, leveraging its flexibility to implement permission management, highly-optimized storage operations, native governance, and modular support.

In short, CDS trades initial complexity for **long-term efficiency, safety, and flexibility**—a tradeoff that is more than justified for systems requiring cutting-edge upgradeability. As audited, plug-and-play implementations become available, CDS will become an **accessible and powerful tool** for developers building systems that demand efficiency and capabilities beyond what is currently accessible.

**Why Use CDS When Diamond Storage Exists?**

> The **Diamond Standard ([ERC-2535](./eip-2535.md))** modularizes smart contract logic elegantly, but retains a rigid storage model. While facets can be upgraded post-deployment, **the storage layout itself remains static**. To mirror the capabilities of CDS, a Diamond storage layer would require its own proxy-delegate solution up-front.  Developers would need to swap the diamond delegate, then re-connect all of its linked contracts.  *In a production environment, this process represents both cost and risk.*  Additionally, this approach further exacerbates the *spiderweb contract structure*.

Reviewer: This subsection should be in your Motivation section. Rationale should be used to explain choices made within your document, while the Motivation justifies the document as a whole.

		    
- **Entry Creation**
    - Input: Add 10,000 entries to a storage space with `pushMany`.
    	- Expected: System can store to  any of these entry indices.
    
- **Invalid Input**
    - Input: `put(42, memberIndex = 1, storageSpace = 0)`.
    	- Expected: Reverts with error.
    
	* Input: `put("42", memberIndex = 1, entryIndex = 0, storageSpace = 0)`.
    	- Expected: Reverts with error.
    
	- Input: `put_string(42, memberIndex = 1, entryIndex = 0, storageSpace = 0)`.
    	- Expected: Reverts with error.

## Gas Benchmarks

* This section assumes that storage operations interact with pre-populated slots.  
* These values are derived from the optimized HoneyBadger model and serve as good targets for efficiency.  
* Displayed values reference *execution* cost.  

	`init_create([1], [256]):` 93,860 gas  
	`init_create([1,1,1], [8,256,256]):` 140,024 gas  
	`insert_new_member:` 40,653 gas  
	`push:` 10,316 gas  
	`put:` 15,543 gas  
	`put_batch([20, 20], [0, 1], 0, 0):` 22,895 gas  
	`get`: 9374 gas  

## Reference Implementation

> Refer to [CDS Minimal Example](https://github.com/wisecameron/ConsolidatedDynamicStorage/blob/main/contracts/CDSMinimal.sol)

Reviewer: Please either include your reference implementation in the `assets/...` directory or remove this section. We do not permit external links except for those listed in EIP-1.

Author: Thank you for taking the time to dig through and find these issues -- will fix each as soon as possible.

---
eip: 7844
title: Consolidated Dynamic Storage (CDS)

> description: Unbounded in-place storage structure upgradeability managed at the contract level.

Reviewer: ```suggestion
description: Unbounded in-place storage structure with upgradeability managed at the contract level.
```

Is this more correct? I'm not sure I understand the original sentence.

Author: It's a word salad for sure.  The premise is that it abstracts away manual storage management entirely by allowing people to extend or create mapped storage structures through function invocations (ie; insert_new_member, init_create).  Will revise this.

author: Cameron Warnick (@wisecameron)
discussions-to: https://ethereum-magicians.org/t/erc-7844-consolidated-dynamic-storage-cds/22217
status: Draft
type: Standards Track
category: ERC
created: 2024-12-13
---

## Abstract  

> **Consolidated Dynamic Storage (CDS)** introduces a dedicated storage contract that provides dynamically extendable, hash-indexed storage spaces shared across multiple contracts. By consolidating structured storage into a single, logically segmented layer, CDS eliminates slot conflicts, enables deterministic schema expansion, and enforces strict separation between storage and execution.  Consolidated multi-contract storage management is also offered by the Diamond Pattern.  CDS expands on this concept by introducing consistent cross-contract access and modification procedures, allowing cross-contract dependencies to leverage standardized interfaces that remain stable regardless of logical or data structure updates.  

Reviewer: ```suggestion
**Consolidated Dynamic Storage (CDS)** introduces a dedicated storage contract that provides dynamically extendable, hash-indexed storage spaces shared across multiple contracts. By consolidating structured storage into a single, logically segmented layer, CDS eliminates slot conflicts, enables deterministic schema expansion, and enforces strict separation between storage and execution.  Consolidated multi-contract storage management is also offered by [ERC-2535](./eip-2535.md)'s Diamond Pattern.  CDS expands on this concept by introducing consistent cross-contract access and modification procedures, allowing cross-contract dependencies to leverage standardized interfaces that remain stable regardless of logical or data structure updates.  
```

6.  **Long-Term Technical Debt and Maintainability** 
	Factors including adapters, siloed storage, multi-hops, and dependent interfaces introduce new risks and considerations, increasing the complexity of upgrades as ecosystem scope expands.  The absence of these considerations streamlines the upgrade process considerably, allowing developers to prioritize optimizing and securing critical business logic.  Additionally, increased lines of code and complexity typically garner higher audit costs, while increasing the odds that vulnerable code makes it to production.

## Specification 

![Adaptability](../assets/erc-7844/Adaptability.svg)

The keywords “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY” and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

> 

Reviewer: I think you could do with a bit more introductory text before jumping straight into requirements. I'm having a bit of a hard time putting the specification into context (though admittedly I'm not a smart contract developer).

Author: Awesome, yeah considering that a lot of the rationale also needs to be moved, I'll do a significant revision and present the narrative in a more straightforward way.  When I initially made this, I was focused too much on the benefits of upgrading storage without re-deployment, when the more fundamental benefit behind all of this is that it's a standardized, holistic approach to storage management in general.  Tertiary benefits like in-place upgradeability, standardized access patterns, etc are all within that fundamental basis, but I don't really communicate this in the proposal.

### Extendable Structs and Storage Spaces

> - Extendable structs **MUST** leverage dynamic mappings with deterministic field hashes (`keccak256`).  

Reviewer: Don't bold the uppercase keywords. Just leave them uppercase, but otherwise unadorned.

- The base struct and its additions **MUST** remain immutable, while fields **MUST** be dynamically appendable.
- Storage spaces **MUST** equate to a simple extension of the hashing structure, segmenting both struct-defining and active data.

Each struct member **MUST** be defined using the following compact metadata structures:

| **Field**  | **Bits** | **Description**                          |
| ---------- | -------- | ---------------------------------------- |
| `bitCount` | 128      | Starting bit offset for the member.      |
| `size`     | 64       | Size of the member in bits.              |
| `type`     | 64       | Type ID (e.g., `uint256`, `bool`, etc.). |

> ### Type IDs:

Reviewer: ```suggestion
### Type IDs
```

| **Type**                   | **ID** | **Size**     |
| -------------------------- | ------ | ------------ |
| `uint`                     | 1      | 8...256 bits |
| `int`                      | 2      | 8...256 bits |
| `bool`                     | 3      | 8 bits       |
| `address`                  | 4      | 160 bits     |
| `bytes32 (optional)`       | 5      | 256 bits     |
| `string (≡ bytes)`         | 6      | Dynamic      |

*For arrays, developers **MAY** define unpacking logic to treat `string` or `bytes` fields as indexed collections of dynamic elements.*

> ### Hash Structure:

Reviewer: ```suggestion
### Hash Structure
```

There are two special values that **MUST** be included: a `safeIndex` and a `stringIndex`.  Both are expounded upon in detail below.

Developers **SHOULD** include a unique hash offset for each space, as this simplifies the development and audit process considerably, and thereby greatly reduces the risk of improper implementation.

A `storageSpace` `offset` value **MUST** be included.  This is explicitly marked in the below section for clarity.  

*MEMBERS_LIM* refers to the total max quantity of struct members permitted within a single extendible struct.  
*ENTRIES_LIM* refers to the max amount of entries in a storage space, and serves as the upper limit for *push* and *pushMany*.  

`A`, `B`, `C` hash offset structure not only allows CDS to functionally guarantee safety from collisions, but also aid in debugging by 
allowing maintainers to instantly track the high-level objective of complex storage operations.

> **Metadata Root:**  

Reviewer: ```suggestion
#### Metadata Root
```

Don't be afraid to use more heading levels.

> **Metadata Root:**  

Reviewer: Re: introductory text, you definitely need some here. You're jumping straight into Yul.

   uint256 entryIndex, 
   uint256 storageSpace
) external;

function get_batch(
    uint256[] memory members, 
    uint256 entryIndex, 
    uint256 storageSpace
) external view returns(uint256[] memory result);

function get_index_from_address(
	address indexAddress,
	uint256 memberIndex,
	uint256 storageSpace
) external view returns(uint256 entryIndex);

function map_address_to_index(
	address indexAddress,
	uint256 storageSpace,
	uint256 targetIndex
) internal;

function put_with_address(
	address indexAddress,
	uint256 storageSpace,
) external;

function _get_root_slot(
	uint256 storageSpace
) internal view returns(uint256);

```
### Initialization

![Basic Interaction Flow](../assets/erc-7844/BasicInteractionFlow.svg)
![CDS Init, Extension](../assets/erc-7844/CDSInitExpansion.svg)  
The `init_create` function **MUST** handle storage space creation in both an initialization and live extension setting.  The function **MUST** take an array of types and sizes as input, which **MUST** conform to the above specifications, and be equivalent in length when input.  Developers **MUST** include length equivalency validation for `types` and `sizes`.

```solidity

> function init_create(

Reviewer: You should move implementations into the reference implementation section (unless there is _exactly_ one possible implementation). The specification section should define the observable behavior of an implementation, not the internals of how it works.

One of the key improvements CDS introduces is standardized storage access across contracts. Whereas Proxy-Delegate and Diamond models require each contract to define its own storage mappings, CDS provides a shared, structured interface that multiple contracts can access consistently. This reduces cross-contract storage fragmentation, simplifies state synchronization, and eliminates the inefficiencies of multi-hop storage dependencies and redundant mappings.

Furthermore, CDS is designed with long-term system maintainability in mind. As protocols expand, managing state across interdependent contracts becomes increasingly complex. CDS mitigates these risks by externalizing structured storage into a permissioned, extendable layer, ensuring that high-throughput contract interactions remain efficient even as ecosystems grow in scale and complexity.

> Proxy-Delegate and Diamond architectures remain valuable for many upgradeable contract designs, offering modularity and logic flexibility. However, for protocols requiring sustainable long-term upgradeability, efficient cross-contract storage, and deterministic schema evolution, CDS provides a more scalable and future-proof alternative. By eliminating migration risks and reducing upgrade complexity, CDS ensures storage integrity, seamless interoperability, and efficient state management—empowering developers to focus on core logic rather than maintenance overhead.

Reviewer: Almost all of this belongs in the Motivation section, because it's justifying the proposal as a whole. The Rationale section is for explaining choices made within the document itself. My favorite analogy is:

> Motivation: We need to build a shed because...
> Rationale: We chose to paint the shed red because...

---
eip: 7844
title: Consolidated Dynamic Storage (CDS)
description: Storage abstraction layer enabling schema-safe upgrades via extendable structs and expandable storage spaces with standardized access patterns.
author: Cameron Warnick (@wisecameron)
discussions-to: https://ethereum-magicians.org/t/erc-7844-consolidated-dynamic-storage-cds/22217
status: Draft
type: Standards Track
category: ERC
created: 2024-12-13
---

## Abstract  

> **Consolidated Dynamic Storage (CDS)** introduces a new foundational storage primitive with broad implications for modern smart contract development.  Its aim is twofold: make storage upgradeability safe and consistent by introducing the concept of *storage abstraction,* and provide a holistic, ready-to-use foundation that naturally fosters good development hygeine, while giving developers space to zero-in on business logic.  It differs considerably from alternative upgradeability solutions such as UUPS and Diamond Storage: CDS is a complete storage management layer, rather than a design pattern.

Author: Yeah, definitely better.  Thanks :)

CDS enables in-place storage structure extension, referred to as *storage abstraction,* introducing mapped storage spaces populated by "extendible structs," and thereby removing any need for storage-driven contract redeployments or manual slot management.  Storage spaces can be dynamically allocated in-place, and are logically segmented using an efficient bit-packed hash structure.  By leveraging basic access controls, development teams can ensure secure storage management for an arbitrary-length collection of linked contracts with varying permissions, gaining the resilience of siloed storage within a unified storage context.  This holistic approach to deterministic storage management facilitates safe upgrades, standardizes storage operations, and enhances interoperability while reducing operational costs and complexity. With on-chain storage considerations managed by CDS, developers are free to focus solely on contract logic, enabling streamlined development and reducing audit scope, while minimizing the odds of vulnerabilities reaching production, especially via upgrades.

> Additionally, CDS optimizes execution efficiency by enforcing a single-hop storage access model, replacing deep delegation chains with direct, permissioned storage interactions. This reduces redundant SLOAD operations and can lower gas overhead in multi-contract systems. These properties make CDS particularly well-suited for DeFi protocols, DAOs, and modular architectures requiring long-term upgradeability without increasing complexity or risk.

Reviewer: You're gonna need to trim down your abstract a bit. There's a lot of good content here. You certainly give a technical overview of the proposal (which is a requirement), but a good abstract also needs to be concise.

Author: Thank you so much for revisiting the proposal, seriously I really appreciate your time and feedback. Really cool work on EELS by the way, it's been pretty useful for me.
